pkg/target/regolib/src.rego

```diff
@@ -200,6 +200,10 @@ matches_nsselector(match) {
 	not has_field(match, "namespaceSelector")
 }
 
+matches_nsselector(match) {
+	has_field(match, "namespaceSelector")
+	not {{.DataRoot}}.cluster["v1"]["Namespace"]
```

Comment:
What is the `not` statement intended to do? It looks like `matches_nsselector` would always match if namespaces aren't cached, which would cause all policies to be applied to all namespaces.

I think we need to handle/auto-reject on the case where the namespace doesn't exist to satisfy the selector logic.

Reply:
The `not` here ensured we catch empty `{{.DataRoot}}.cluster["v1"]["Namespace"]` first before `{{.DataRoot}}.cluster["v1"]["Namespace"][input.review.namespace]` silently fails in the following `matches_nsselector`. 

```go
matches_nsselector(match) {
	has_field(match, "namespaceSelector")
	ns := {{.DataRoot}}.cluster["v1"]["Namespace"][input.review.namespace]
	matches_namespace_selector(match, ns)
}
```
> I think we need to handle/auto-reject on the case where the namespace doesn't exist
This implementation has the same effect as if `namespaceSelector` was not defined in the constraint, ie auto-reject across all namespaces if the resource failed to satisfy the constraint.

Reply:
`not` could be replaced w/a `get_default` and some logic to return a rejection if the namespace isn't in the cache.

WRT "This implementation has the same effect as if namespaceSelector was not defined in the constraint", I don't think this is an ideal solution. It may be more confusing than helpful.

i.e. "why is this constraint that is only supposed to apply to prod namespaces applying to my dev namespace? Everything is configured correctly"

Reply:
I thought this implements what option 1 was from #153. 
> Deny requests when a namespace selector runs up against a non-existent namespace

maybe I misunderstood. If you want to reject based on the namespace in the request, then we would have to inject the namespace somehow. then that would be option 2.

Reply:
Deny requests, not apply all constraints to those requests. This means that if a namespace is not yet cached, we auto-fail. If we applied all constraints, there would still be a chance of acceptance.

Reply:
I think I see the issue. The namespace name is still listed in `input.review.namespace`, even when it hasn't been cached yet.

Reply:
Right, the `{{.DataRoot}}.cluster["v1"]["Namespace"]` cache is empty when we dont sync namespace, but we will always get `input.review.namespace` from the request.

Reply:
So there should be enough info for a full rejection: Namespace selector + namespace name + no cached namespace -> auto rejection with explanation.

Reply:
> Deny requests, not apply all constraints to those requests. This means that if a namespace is not yet cached, we auto-fail. If we applied all constraints, there would still be a chance of acceptance.

So even if this example https://github.com/open-policy-agent/gatekeeper/example/resources/bad_pod_namespaceselector.yaml had the right `gatekeeper` label, you still want the request to fail as long as we did not cache the namespace?

Reply:
Ya